src/ts/proxyfowarder.ts

```diff
@@ -0,0 +1,159 @@
+import { keccak256, defaultAbiCoder, solidityKeccak256 } from "ethers/utils";
+import { Wallet } from "ethers/wallet";
+import { ReplayProtectionAuthority } from "./replayprotectionauthority";
+import { ChainID, ProxyAccountDeployer, ProxyAccountFactory } from "..";
+import { ForwardParams, Forwarder, ProxyCallData } from "./forwarder";
+
+/**
+ * A single library for approving meta-transactions and its associated
+ * replay protection. All meta-transactions are sent via proxy contracts.
+ */
+export class ProxyForwarder extends Forwarder<ProxyCallData> {
+  /**
+   * All meta-transactions are sent via an proxy contract.
+   * @param chainID Chain ID
+   * @param proxyHub Address of contract
+   * @param signer Signer's wallet
+   * @param replayProtectionAuthority Extends implementation ReplayProtectionAuthority
+   */
+  constructor(
+    chainID: ChainID,
+    proxyHub: ProxyAccountDeployer,
+    signer: Wallet,
+    replayProtectionAuthority: ReplayProtectionAuthority
+  ) {
+    super(chainID, proxyHub, signer, replayProtectionAuthority);
+  }
+
+  /**
+   * Standard encoding for contract call data
+   * @param data The target contract, value (wei) to send, and the calldata to execute in the target contract
+   */
+  protected getEncodedCallData(data: ProxyCallData) {
+    // ProxyAccounts have a "value" field.
+    return defaultAbiCoder.encode(
+      ["address", "uint", "bytes"],
+      [data.target, data.value, data.callData]
+    );
+  }
+
+  /**
+   * Returns the encoded calldata for creating a proxy contract
+   * No need for ForwardParams as no signature is required in ProxyAccountDeployer
+   */
+  public async createProxyContract() {
+    const deployed = await this.forwarder
+      .connect(this.signer)
+      .accounts(this.signer.address);
+
+    // Does the user have a proxy contract?
+    if (deployed === "0x0000000000000000000000000000000000000000") {
+      const callData = this.forwarder.interface.functions.createProxyAccount.encode(
+        [this.signer.address]
+      );
+
+      return callData;
+    }
+
+    throw new Error(
+      "ProxyAccount for " + this.signer.address + " already exists."
+    );
+  }
+
+  /**
+   * ProxyAccount address for this signer
+   * Caution: Contract may not be deployed yet.
+   */
+  public async getProxyAddress() {
+    const baseAddress = await this.forwarder.baseAccount();
```

Comment:
Probably better to memoize this property via a property getter, so as not to keep refetching from chain.

Reply:
I changed so it'll fetch the first time and then keep it around.

Reply:
yeah, I guess it's fine to do it this way - but it's generally nicer to put it into a function so that many other functions can use baseAddress if they want to without thinking about memoization. Eg
```
private mBaseAccount;
private async getBaseAccount() {
    if(!mBaseAccount) {
         this.mBaseAccount = await new ProxyAccountFactory(this.signer).attach(this.proxyAccountFactoryAddress).baseAccount();
    }
    return this.mBaseAccount;
}

```

Reply:
done